rust/impl/ca/ca/context/programstate.rs

```diff
@@ -0,0 +1,64 @@
+
+pub( crate ) mod private
+{
+  /// Execution statement of a program
+  #[ derive( Debug, Default ) ]
+  pub struct ProgramState
+  {
+    /// Current instruction number
+    current_pos : usize
+  }
+
+  impl ProgramState
+  {
+    /// Sets current instruction
+    pub fn set_pos( &mut self, pos : usize )
+    {
+      self.current_pos = pos;
+    }
+
+    /// Returns the number of next instruction
+    pub fn next( &self ) -> usize
+    {
+      self.current_pos + 1
+    }
+
+    /// Sets current instruction to next one
+    pub fn go_next( &mut self )
+    {
+      self.current_pos += 1
+    }
+
+    /// Returns the number of previous instruction
+    pub fn prev( &self ) -> usize
+    {
+      self.current_pos - 1
+    }
+
+    /// Sets current instruction to previous one
+    pub fn go_back( &mut self )
+    {
+      self.current_pos -= 1
+    }
+
+    /// Get number of current instruction
+    pub fn get_pos( &self ) -> usize
+    {
+      self.current_pos
+    }
+
+    /// Sets current instruction to end of program
+    /// Last instruction will not be executed
+    pub fn finish( &mut self )
+    {
+      self.set_pos( usize::MAX )
```

Comment:
Is it safe? I mean access to element with index `usize::MAX`.

Reply:
Hmm, it looks like it realy isn't safe when a program has usize::MAX commands.

```
while let Some( instruction ) = instructions.get( state.current_pos )
{
     state.current_pos += 1; // here
     self._instruction_perform( instruction )?;
}
```

But what's the best way to fix it?

Reply:
The best way to solve the issue is using the Option/Result as return type of `set_pos`. You can handle the exception or propagate the error.